### PR TITLE
Release 4.9.0

### PR DESCRIPTION
**Pull request by nim65s**
Hi,

I'm almost done on the 4.9.0 release.

This PR wiill be the main place to discuss it.

I mostly  took master branches of most HPP packages, plus devel branches of the HPP-RBPRM tree, and set up CMake Exports. For now, .pc files are still created, but not used, and therefore not tested.

This is only a first step, after that we will probably be able to ensure that all packages are [relocatables](https://cmake.org/cmake/help/v3.3/manual/cmake-packages.7.html#creating-relocatable-packages), and get rid of some legacy stuff.

There is also one annoying thing, already discussed in https://github.com/jrl-umi3218/jrl-cmakemodules/pull/353: the configure step is currently actually looking several times for some dependencies of dependencies.

If you need anything included or removed from this release, this is still possible, please tell me :)

:warning: **Please don't merge those pull requests** :warning: : they are still a work in progress, will be modified (with a new version of the jrl-cmakemodule with https://github.com/jrl-umi3218/jrl-cmakemodules/pull/371, at least), and merging those would break the tags I have when I already have them.

On finalizing this release, I will merge the v4.9.0 tags to master, and then master to devel, in case there already are some commits in devel that should not be included in 4.9.0 (which is already the case in several places), while ensuring that master and devel did not diverge.

### CODE REVIEW

**Comment by nim65s** [comment]
Here is a list of things that changed, so we can discuss those more easily:

1. all variables that were written in the generated .pc are now also in the generated .cmake
2. use (1) implies update eg. `HPP_PACKAGE_FOUND` to `hpp-package_FOUND`
3. main `ADD_LIBRARIES` have been moved from `src/CMakeLists.txt` to the main `CMakeLists.txt`, to easily set its sources as `${PROJECT_NAME}_SOURCES}` and `${PROJECT_NAME}_HEADERS`. The header list was already defined in the main `CMakeLists.txt`, and I think it's useful to add those to the library, so that CMake knows that on an update of one of those headers, the main lib must be updated. It might also be useful for `ADD_SOURCE_GROUP` & `ADD_HEADER_GROUP` for IDEs, but I didn't test that yet.
4. if a package depend on A and B, and A already depends on B, I removed the dependency to B, as the generated .cmake of A already checks for B. This helps for the "annoying thing" I mentioned earlier, but is not enough to fix it, because of diamond dependencies.
5. I removed some dependencies that are not build-time dependencies, and that are not necessary for any test. Eg. hpp-baxter was depending on hpp-corbaserver, because it installs python files in `$PYTHON_SITELIB/hpp/corbaserver/…`, but actually we can install those files anyway, and as it also install other files (robot description) that are required for hpp-core (which is a dependency of hpp-corbaserver), this was usefull to break a dependency cycle.
6. Added some interface libraries (`ADD_LIBRARY(${PROJECT_NAME} INTERFACE)`) for projects that only install plain files: this allows to generate a .cmake, which is useful to check if those files are installed or not in a dependent package
7. I moved cmake files from `share/${PROJECT_NAME}/cmake/` to `lib/cmake/${PROJECT_NAME}`, because this is the place where the generated .cmake are going, so I think it is better to have them all in the same place. According to the [CMake doc](https://cmake.org/cmake/help/latest/command/find_package.html#search-procedure), the first unix path in the search list is `<prefix>/(lib/<arch>|lib*|share)/cmake/<name>*/ `, while `<prefix>/(lib/<arch>|lib*|share)/<name>*/(cmake|CMake)/` is the last.
8. I removed the `ADD_DOC_DEPENDENCY`. This was by laziness, because I didn't knew exactly what that did, but guessed that this was using the generated .pc. I will investigate more before the release.
9. I had to add some `INCLUDE(${GEPETTO_VIEWER_CMAKE_PLUGIN})`, but this is probably not required, and would require further investigations too, but that can probably wait a bit.

**Comment by jmirabel** [comment]
>     8\. I removed the `ADD_DOC_DEPENDENCY`. This was by laziness, because I didn't knew exactly what that did, but guessed that this was using the generated .pc. I will investigate more before the release.

I think this is only to get the doxytag working.



> * 
> 
> * I had to add some `INCLUDE(${GEPETTO_VIEWER_CMAKE_PLUGIN})`, but this is probably not required, and would require further investigations too, but that can probably wait a bit.

I think the `include` can be added to the generated `config.cmake`, no ?

**Comment by nim65s** [comment]
For the doc, I tried to remove
https://github.com/humanoid-path-planner/hpp-core/blob/dda9aa14b8bf9a1a787353addf51fa15816aaf2f/CMakeLists.txt#L146

And if I go into the generated doc, and type pinocchio in the search bar, I'm successfully redirected to my file:///usr/share/doc/pinocchio/doxygen-html/namespacepinocchio.html ; is this the behavior you are expecting ? Is this enough to test that ? Or should I look for anything else ?

For the include, yes, that was the solution, and it works :)

**Comment by jmirabel** [comment]
> Or should I look for anything else ?

Your check is sufficient. Alternatively, you can simply check that *TAGFILES* in *build/doc/Doxyfile* (so after you ran `make doc`) contains *pinocchio* (and not only *hpp-pinocchio*).

If the *...Config.cmake* was calling *ADD_PROJECT_DEPENDENCY*, then there would be nothing to do.
It does not and if we don't want this (which I can understand), we can modify the *...Config.cmake* in the cmake module so that it fills:
- either it fills `_PACKAGE_CONFIG_DEPENDENCIES_PROJECTS` for relevant `find_package`,
- or it fills a new variable `_PACKAGE_DEPENDENCIES_DOXYTAG`, that has to be read in *doxygen.cmake* (macro `_SETUP_PROJECT_DOCUMENTATION_FINALIZE`).

**Comment by nim65s** [comment]
Thanks. I checked the generated Doxyfile, and it contains `/usr/share/doc/pinocchio/doxygen-html/pinocchio.doxytag`, so I guess the dependency chain in the generated …Config.cmake is using the right `FIND_PACKAGES` (or our `ADD_PROJECT_DEPENDENCY` wrapper).

Then I guess we are pretty much ready for this release :)

File: src/path-optimization/simple-time-parameterization.cc [hunk 1]
```diff
@@ -30,242 +30,213 @@
...
-            "SimpleTimeParameterization/maxAcceleration",
-            "The maximum acceleration for each degree of freedom. Not considered if negative.",
-            Parameter((value_type)-1)));
-      HPP_END_PARAMETER_DECLARATION(SimpleTimeParameterization)
-    } // namespace pathOptimization
-  } // namespace core
-} // namespace hpp
+namespace core {
+namespace pathOptimization {
+using std::isfinite;
```
Comment (jmirabel) [inline src/path-optimization/simple-time-parameterization.cc:35]: I would prefer to split apart re-indentation and actual changes.

Comment (nim65s) [reply src/path-optimization/simple-time-parameterization.cc:35]: Eeek, I didn't even see that, thanks.

**Pull request merged**